tests/test_image.py

```diff
@@ -1,233 +1,282 @@
-        shape = (*self.mdim_ims_np.shape[:-2][::-1], *self.mdim_ims_np.shape[-2:])
-        self.assertTrue(np.allclose(X.asnumpy(), self.mdim_ims_np.reshape(shape)))
-
-    def testMultiDimFlattens(self):
-        # Try flattening
-        X = self.mdim_ims.stack_reshape(self.mdim_ims.n_images)
-        self.assertEqual(X.stack_shape, (self.mdim_ims.n_images,))
-
-    def testMultiDimFlattensTrick(self):
-        # Try flattening with -1
-        X = self.mdim_ims.stack_reshape(-1)
-        self.assertEqual(X.stack_shape, (self.mdim_ims.n_images,))
-
-    def testMultiDimReshapeTuples(self):
-        # Try flattening with (-1,)
-        X = self.mdim_ims.stack_reshape((-1,))
-        self.assertEqual(X.stack_shape, (self.mdim_ims.n_images,))
-
-        # Try mdim reshape
-        X = self.mdim_ims.stack_reshape(self.mdim_ims.stack_shape[::-1])
-        self.assertEqual(X.stack_shape, self.mdim_ims.stack_shape[::-1])
-
-    def testMultiDimBadReshape(self):
-        # Incorrect flat shape
-        with self.assertRaisesRegex(ValueError, "Number of images"):
-            _ = self.mdim_ims.stack_reshape(8675309)
-
-        # Incorrect mdin shape
-        with self.assertRaisesRegex(ValueError, "Number of images"):
-            _ = self.mdim_ims.stack_reshape(42, 8675309)
-
-    def testMultiDimBroadcast(self):
-        X = self.mdim_ims + self.ims
-        self.assertTrue(np.allclose(X[0], 2 * self.ims.asnumpy()))
+n = 3
+mdim = 2
+
+
+def get_images(parity=0, dtype=np.float32):
+    size = 768 - parity
+    # numpy array for top-level functions that directly expect it
+    im_np = misc.face(gray=True).astype(dtype)[np.newaxis, :size, :size]
+    denom = np.max(np.abs(im_np))
+    im_np = im_np / denom  # Normalize test image data to 0,1
+    # Independent Image object for testing Image methods
+    im = Image(im_np.copy())
+    return im_np, im
+
+
+def get_stacks(parity=0, dtype=np.float32):
+    im_np, im = get_images(parity, dtype)
+
+    # Construct a simple stack of Images
+    ims_np = np.empty((n, *im_np.shape[1:]), dtype=dtype)
+    for i in range(n):
+        ims_np[i] = im_np * (i + 1) / float(n)
+
+    # Independent Image stack object for testing Image methods
+    ims = Image(ims_np)
+
+    return ims_np, ims
+
+
+def get_mdim_images(parity=0, dtype=np.float32):
+    ims_np, im = get_stacks(parity, dtype)
+    # Multi dimensional stack Image object
+    mdim = 2
+    mdim_ims_np = np.concatenate([ims_np] * mdim).reshape(mdim, *ims_np.shape)
+    mdim_ims = Image(mdim_ims_np)
+
+    return mdim_ims_np, mdim_ims
+
+
+def testRepr():
+    _, mdim_ims = get_mdim_images()
+    r = repr(mdim_ims)
+    logger.info(f"Image repr:\n{r}")
+
+
+def testNonSquare():
+    # don't need to parametrize this test
+    """Test that an irregular Image array raises."""
+    with raises(ValueError, match=r".* square .*"):
+        _ = Image(np.empty((4, 5)))
+
+
+@pytest.mark.parametrize("parity,dtype", params)
```

Comment:
No space after the comma in `"parity,dtype"` really bugs me, but I see that's the convention in pytest docs.